Enterprise API Usage Report

jingcheng16 opened the pull request:
## Product Description
<!-- Where applicable, describe user-facing effects and include screenshots. -->
This PR is to add Enterprise API Usage report in enterprise console. This tile will live in Security Center.

**UI:**
<img width="1447" alt="image" src="https://github.com/user-attachments/assets/2fe0719f-05e1-41a1-a7af-13e0ebd408a8" />

**The downloaded report:**
<img width="569" alt="image" src="https://github.com/user-attachments/assets/5dcf75d6-7786-42e9-90eb-e6abb12379d4" />

**API**
```shell
curl -s -X GET "http://localhost:8000/a/enterpriseconsoletest/enterprise/api/v1/apiusage/" -H "Authorization: ApiKey xxx:xxxxxxxxxxxxx"
{"@odata.context": "http://localhost:8000/a/enterpriseconsoletest/enterprise/api/v1/apiusage/schema/#feed",
"@odata.count": 5,
"value": [{"api_key_name": "sdfs", "created_date": "2024-01-23T12:58:40+00:00", "expiration_date": null, "last_used_date": null, "scope": "All project spaces", "web_user": "x@dimagi.com"},
{"api_key_name": "Test", "created_date": "2024-07-02T20:13:03+00:00", "expiration_date": null, "last_used_date": "2024-07-07T18:18:58+00:00", "scope": "All project spaces", "web_user": "x@dimagi.com"},
{"api_key_name": "limited", "created_date": "2024-11-13T22:42:09+00:00", "expiration_date": null, "last_used_date": null, "scope": "testproject1", "web_user": "x@dimagi.com"},
{"api_key_name": "For Enterprise Tile", "created_date": "2024-12-16T05:22:14+00:00", "expiration_date": "2024-12-31T00:00:00+00:00", "last_used_date": null, "scope": "inconsistent1", "web_user": "x@dimagi.com"},
{"api_key_name": "locallll", "created_date": "2024-09-04T15:06:07+00:00", "expiration_date": null, "last_used_date": "2024-12-17T21:56:25+00:00", "scope": "All project spaces", "web_user": "x@dimagi.com"}]}%
``` 

## Technical Summary
<!--
    Provide a link to any tickets, design documents, and/or technical specifications
    associated with this change. Describe the rationale and design decisions.
-->
Ticket: https://dimagi.atlassian.net/browse/SAAS-16378

The implementation is straight forward. All the information required in this report lives in `HQApiKey`.

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->
Tested locally.

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->


### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

jingcheng16 committed:
Add API Usage Tile

jingcheng16 committed:
add Api Usage api

mjriley commented on corehq/apps/enterprise/enterprise.py:
```diff
@@ -561,3 +564,36 @@ def rows_for_domain(self, domain_obj):
+        return self.unique_api_keys().count()
+
+    def unique_api_keys(self):
+        usernames = self.account.get_web_user_usernames()
+        unique_users = [User.objects.get(username=username) for username in usernames]
+        return HQApiKey.objects.filter(user__in=unique_users, is_active=True)
```
Three things we might want to look into here:
1. Can we simplify what is pulled back from the database -- i.e. instead of fetching the entire user object, can we fetch just the user ID or username
2. Can/should this be turned into a subquery so that we only send a single query to the database?
3. At what # of users are we concerned with `__in` performance? Maybe run this by the infrastructure team.

mjriley commented on corehq/apps/enterprise/enterprise.py:
```diff
@@ -561,3 +564,36 @@ def rows_for_domain(self, domain_obj):
+        return [
+            api_key.user.username,
+            api_key.name,
+            api_key.domain if api_key.domain else 'All project spaces',
```
Translate "All project spaces"

jingcheng16 committed:
Simplify the query and use Subquery

jingcheng16 committed:
Exclude api key whose scope is not to any project spaces in this billing account

jingcheng16 committed:
Reformat the report row

Although some api key's scope is all project spaces, but we should only show project space that is a member of the billing account and the user is a member of, which reflects the actual scope

jingcheng16 committed:
Merge branch 'master' into jc/api-usage-enterprise-report
